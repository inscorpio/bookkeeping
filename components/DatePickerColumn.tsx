import useTouchMove from '~/hooks/useTouchMove'

interface Props {
  columns: string[]
  selectIndex: number
  onSelect: (selectIndex: number) => void
}

export const ITEM_HEIGHT = 28

// !!! 注意 translateY 是负数
export default function DatePickerColumn({ columns, selectIndex, onSelect }: Props) {
  const initialMovedIstance = selectIndex * ITEM_HEIGHT
  const { moveDistance, touchStart, touchMove } = useTouchMove(selectIndex, columns.length - 1)
  const finalDistance = initialMovedIstance + moveDistance
  const handleTouchEnd = () => {
    onSelect(finalDistance / ITEM_HEIGHT)
  }

  return (
    <li
      className="flex flex-col items-center overflow-hidden"
      onTouchStart={touchStart}
      onTouchMove={touchMove}
      onTouchEnd={handleTouchEnd}
    >
      {
        columns.map((v, i) => {
          return (
            <span
              key={`${v}-${i}`}
              style={{ transform: `translateY(${(finalDistance) * -1}px)` }}
            >
              {v}
            </span>
          )
        })
      }
    </li>
  )
}
