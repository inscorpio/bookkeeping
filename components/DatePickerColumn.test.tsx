import { fireEvent, render, screen } from '@testing-library/react'
import { expect, test } from 'vitest'
import DatePickerColumn, { ITEM_HEIGHT } from '~/components/DatePickerColumn'
import { touchMoveDown, touchMoveUp } from '~/components/DatePickerColumn.helper'

const columns = Array(10).fill('')

test('render', () => {
  const { container } = render(
    <DatePickerColumn
      columns={columns}
      selectIndex={0}
      onSelect={() => {}}
    />,
  )

  expect(container).toMatchSnapshot()
})

test('basic', async () => {
  let selectIndex = 0
  render(
    <DatePickerColumn
      columns={columns}
      selectIndex={selectIndex}
      onSelect={(index) => {
        selectIndex = index
      }}
    />,
  )

  const li = screen.getByRole('listitem')

  await touchMoveUp(li)

  expect(li.firstChild).toHaveStyle(`transform: translateY(-${ITEM_HEIGHT}px)`)
  expect(selectIndex).toBe(1)
})

test.todo('移动距离超过一半时 step +1', async () => {
  let selectIndex = 0
  render(
    <DatePickerColumn
      columns={columns}
      selectIndex={selectIndex}
      onSelect={(index) => {
        selectIndex = index
      }}
    />,
  )

  const li = screen.getByRole('listitem')

  await touchMoveUp(li)

  expect(li.firstChild).toHaveStyle(`transform: translateY(-${ITEM_HEIGHT}px)`)
  expect(selectIndex).toBe(1)
})

test('multiple moves', async () => {
  render(
    <DatePickerColumn
      columns={columns}
      selectIndex={0}
      onSelect={() => {}}
    />,
  )
  const li = screen.getByRole('listitem')

  await touchMoveUp(li)

  expect(li.firstChild).toHaveStyle(`transform: translateY(-${ITEM_HEIGHT}px)`)

  await touchMoveUp(li)

  expect(li.firstChild).toHaveStyle(`transform: translateY(-${ITEM_HEIGHT * 2}px)`)
})

test('should not exceed the first position when move down', async () => {
  let selectIndex = 3
  render(
    <DatePickerColumn
      columns={columns}
      selectIndex={selectIndex}
      onSelect={(index) => {
        selectIndex = index
      }}
    />,
  )

  const li = screen.getByRole('listitem')
  const step = 4

  await touchMoveDown(li, {
    step,
  })

  expect(li.firstChild).toHaveStyle(`transform: translateY(0px)`)
  expect(selectIndex).toBe(0)
})

test('should not exceed the last position when move up', async () => {
  const maxStep = columns.length - 1
  const step = 2
  let selectIndex = 8
  render(
    <DatePickerColumn
      columns={columns}
      selectIndex={selectIndex}
      onSelect={(index) => {
        selectIndex = index
      }}
    />,
  )

  const li = screen.getByRole('listitem')

  await touchMoveUp(li, {
    step,
  })

  expect(li.firstChild).toHaveStyle(`transform: translateY(-${ITEM_HEIGHT * maxStep}px)`)
  expect(selectIndex).toBe(maxStep)
})
