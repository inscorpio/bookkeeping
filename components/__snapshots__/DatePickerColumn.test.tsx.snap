// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render 1`] = `
<div>
  <li
    class="flex flex-col items-center overflow-hidden"
  >
    <span
      style="transform: translateY(0px);"
    />
    <span
      style="transform: translateY(0px);"
    />
    <span
      style="transform: translateY(0px);"
    />
    <span
      style="transform: translateY(0px);"
    />
    <span
      style="transform: translateY(0px);"
    />
    <span
      style="transform: translateY(0px);"
    />
    <span
      style="transform: translateY(0px);"
    />
    <span
      style="transform: translateY(0px);"
    />
    <span
      style="transform: translateY(0px);"
    />
    <span
      style="transform: translateY(0px);"
    />
  </li>
</div>
`;
